### DISCUSSION TITLE
Pattern for merging into a default object

### DISCUSSION THREAD

**Discussion by smozely** [remeda/remeda#977] [Q&A] [answered]
I'm trying to merge a partial object into a default object. 

Wondering if there is a correct pattern to get it to end up with the correct non-partial type as the result?

e.g. this errors because resultFoo isn't compatible with FooData
```
interface FooData {
  id: string;
  email: string;
}

const defaultFoo: FooData = {
  id: '123',
  email: 'abc',
};

const partialFoo: Partial<FooData> = {
  id: '321',
  email: 'cba',
};

const resultFoo: FooData = R.merge(defaultFoo, partialFoo);
console.log(resultFoo);
```

And I understand why it does that, partial foo could have something like 
```
const partialFoo: Partial<FooData> = {
  id: '321',
  email: undefined,
};
```

I can work around it by doing:
`const resultFoo = R.merge(defaultFoo, partialFoo) as FooData;`

But just wondering if there are any good patterns to avoid needing to use an `as`

**Comment by smozely** [comment]
Interestingly, and I don't understand why ... this works though (maybe it shouldn't)

```
import * as R from 'remeda';

interface FooData {
  id: string;
  email: string;
}

const defaultFoo: FooData = {
  id: '123',
  email: 'abc',
};

const nested: { partialFoo?: Partial<FooData> } = {
  partialFoo: {
    id: '321',
    email: 'cba',
  },
};

const resultFoo: FooData = R.merge(defaultFoo, nested.partialFoo);
console.log(resultFoo);
```

**Comment by eranhirsch** [comment]
This is a really interesting edge case. I don't understand it myself yet. We rely pretty heavily on the type-fest type `Merge` for `merge`, and the issue can be traced back to this type. The type worked in your second example because it's possibly undefined, and I guess that causes type-fest to go through a different branch. This might be a bug in type-fest.

https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true&noUnusedParameters=true&noPropertyAccessFromIndexSignature=true&noImplicitOverride=true&noFallthroughCasesInSwitch=true&exactOptionalPropertyTypes=true&useUnknownInCatchVariables=true&strictBuiltinIteratorReturn=true&noUnusedLocals=true#code/JYWwDg9gTgLgBAbzgWQKZQOargXzgMyghDgCIYBPMVAWn1QGcZSBuAKDeADsZ18BDAMbYAYhAgARfjH6I2cOMAAmALjhMo3DOwWoQ-YABs1GrexwdK1OAEEAjHAC8KdFgA8YydP4AaOAAV+WGB+Qw9xKRkAPjgAHzgAVy4lVHxuVCUo9gB6bIU4AD0AfksqbBsAJicXTFRwrxk-QODQ+sj+KKy2XPzioA

It's worthwhile trying to ping the type-fest team about it. This might be intentional, and then we might need to work around it to fix it in Remeda.

**Comment by tbui17** [comment, marked as answer]
# Intro
type-fest's core `Merge` implementation involves a `SimpleMerge` type that iterates over the keys of the types with `keyof`. The type extracts out fields that aren't shared with the source type, then combines those with the fields of the source type, emulating overriding behavior.
```typescript
// Merges two objects without worrying about index signatures.
type SimpleMerge<Destination, Source> = {
	[Key in keyof Destination as Key extends keyof Source ? never : Key]: Destination[Key];
} & Source;
```
`Destination` would be `FooData`
`Source` would be `Partial<FooData> | undefined`

# Side note on index signatures
For reference, index signatures are things like 
```typescript
type A = {
    [key: string]: string;
}
```
These aren't relevant to the issue at hand, but it's worth noting that interfaces don't have this implicitly, unlike object types made with the `type` keyword. You may sometimes get unexpected behavior with interfaces related to this. type-fest's `Merge` implementation has separate handling for index signatures.

```typescript
export type Merge<Destination, Source> =
Simplify<
SimpleMerge<PickIndexSignature<Destination>, PickIndexSignature<Source>>
& SimpleMerge<OmitIndexSignature<Destination>, OmitIndexSignature<Source>>
>;
```

# How it works
When you access the optional key, you get `Partial<FooData> | undefined`. Once we pass that and the `FooData` type into `merge`, this is what we will get:
```typescript
type SimpleMergedData = {
	[Key in keyof FooData as Key extends keyof (Partial<FooData> | undefined) ? never : Key]: FooData[Key];
} & (Partial<FooData> | undefined);
```

Things start to get weird when you involve union types, especially with `undefined`. `keyof` gets the keys common to all members of the union type, returning a `never` if there's nothing in common. `Undefined` has no keys in the first place so you'll never get anything in common, the result will always be `never`. This causes the `keyof Source` to become `never`.

```typescript
type SimpleMergedData = {
	[Key in keyof FooData as Key extends never ? never : Key]: FooData[Key];
} & (Partial<FooData> | undefined);
```

This effectively causes the keys and values of the `Destination` to always be included, reconstructing the `FooData` type. What you end up with is

`FooData & (Partial<FooData> | undefined)`

Typescript will distribute the intersection over the union

 `(FooData & Partial<FooData>) | (FooData & undefined)`

For shared fields, their optionality is overridden by non-optionality, and an intersection type is made out of their respective types. Since `Partial<FooData>` has the same types, all the types just form intersection types with themselves which is a no-op. You effectively just get a type equivalent to `FooData`. 

To explain how `FooData & undefined` evaluates to `never`, we would need to break the abstraction and look into how Typescript works with sets of valid values and mathematical set operations. Dan Vanderkam's _Effective Typescript, Second Edition_ in _Item 7: Think of Types as Sets of Values_ explains this well. Long story short, types constrain the set of valid values ( i.e. `Boolean` can be one of `{true, false}` ). The (respective) sets of valid values for the `FooData` and `undefined` types have nothing in common (no mathematical intersection), so we just get `never`. 

Therefore, you end up with the following:

`FooData | never`

Which simplifies to

`FooData // mathematical union of "FooData" and "never" (no valid values) is just "FooData"` 

Therefore, with `Partial<FooData> | undefined`, you get `FooData`. With just `Partial<FooData>`, the type "works as intended" and overrides the non-optional fields of `FooData` with the optional fields of `Partial<FooData>`. You effectively get `Partial<FooData>` which can't be assigned to `FooData`.

# Solutions

A workaround might involve special handling for unions, `undefined`, and optional fields, possibly by using type distribution to isolate edge cases. This would probably be fairly complex to implement.

In the meantime, you could just spread the objects the old-fashioned way which will produce the (mostly) correct type.

```typescript
interface FooData {
  id: string;
  email: string;
}

const defaultFoo: FooData = {
  id: '123',
  email: 'abc',
};

const partialFoo: Partial<FooData> = {
  id: '321',
  email: 'cba',
};

const resultFoo: FooData = {...defaultFoo,...partialFoo};
console.log(resultFoo);
```

Unfortunately, this isn't 100% safe when you have undefined as a possible type.

```typescript
interface FooData {
  id: string;
  email: string;
}

interface FooData2 {
  id: string|undefined
  email: string
}

const defaultFoo: FooData = {
  id: '123',
  email: 'abc',
};

const partialFoo: Partial<FooData2> = {
  id: undefined,
  email: 'cba',
};

const resultFoo: FooData = {...defaultFoo,...partialFoo}; // no compiler error
console.log(resultFoo);
```

```
[LOG]: {
  "id": undefined,
  "email": "cba"
} 
```
